Add linear least squares version of symmetric icp

matthewdcong opened the pull request:
This addresses issue #3276 by implementing the symmetric objective function for ICP proposed by https://gfx.cs.princeton.edu/pubs/Rusinkiewicz_2019_ASO/index.php.

A basic test has been added to validate the functionality on a simple example.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,251 @@
+                             const pcl::PointCloud<PointTarget> &cloud_tgt,
+                             Matrix4 &transformation_matrix) const
+{
+  size_t nr_points = cloud_src.points.size ();
```
Prefer `const std::size_t` (or simply `const auto`) (and same below)

matthewdcong replied:
Changed to const auto

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,251 @@
+                               Matrix4 &transformation_matrix) const
+{
+  // Construct the transformation matrix from rotation and translation 
+  transformation_matrix = Eigen::Matrix<Scalar, 4, 4>::Zero ();
```
Will the Eigen3 functions to create transform from euler angles help here?

matthewdcong replied:
Yup, switched to using those.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,251 @@
+  // Approximate as a linear least squares problem
+  while (source_it.isValid () && target_it.isValid ())
+  {
+    if (!std::isfinite (source_it->x) ||
```
There should be a function for this (there isn't), this looks brittle

matthewdcong replied:
Added an implementation of isFinite for PointNormals and switched to using that instead.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,251 @@
+  ATb.setZero ();
+
+  // Approximate as a linear least squares problem
+  while (source_it.isValid () && target_it.isValid ())
```
A `for` loop would allow you to have the increments in one place

matthewdcong replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,251 @@
+    double b = nx*sz - nz*sx; 
+    double c = ny*sx - nx*sy;
+   
+    //    0  1  2  3  4  5
```
Why not simply use `TriangularView` to show intent and then `(x,y)` to refer to rows and columns?

matthewdcong replied:
Good idea, changed.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,251 @@
+    ++target_it;
+    ++source_it;    
+  }
+  ATA.coeffRef (6) = ATA.coeff (1);
```
A loop to copy might be more terse (or a triangular view to copy, but as per benchmarks, loop is a little faster and comparable to rolled out)

matthewdcong replied:
Switched to a selfadjointView  and Cholesky solve to avoid the copy to the lower triangular half altogether.

kunaltyagi commented on registration/include/pcl/registration/transformation_estimation_symmetric_point_to_plane_lls.h:
```diff
@@ -0,0 +1,154 @@
+          * \param[out] transformation_matrix the resultant transformation matrix
+          */
+        inline void
+        constructTransformationMatrix (const double & alpha, const double & beta, const double & gamma,
```
Why do you prefer a longform function parameter list? For me personally, it makes more sense looking at an Eigen::Vector3d (or double[3]) translation. Same for the euler angles. (Sorry if the angles aren't Euler. Such rotations are usually XYZ Euler, but I haven't read the paper)

matthewdcong replied:
Switched to a Vector6 parameter list which is already being used by the estimateRigidTransformation caller.

kunaltyagi commented on test/registration/test_registration_api.cpp:
```diff
@@ -655,6 +656,55 @@ TEST (PCL, TransformationEstimationPointToPlane)
+{
+  pcl::registration::TransformationEstimationSymmetricPointToPlaneLLS<pcl::PointNormal, pcl::PointNormal> transform_estimator;
+
+  // Create a test cloud
```
What is this test case representing physically?

In case the magic numbers need to be updated in the future, some physical intuition can guide the change.

matthewdcong replied:
Added a comment with the corresponding rotation and translation values.

matthewdcong commented:
Thanks for the feedback, Kunal! I've addressed it in my latest commit.

matthewdcong commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -222,30 +196,12 @@ estimateRigidTransformation (ConstCloudIterator<PointSource>& source_it, ConstCl
 
   // Solve A*x = b
-  Vector6d x = static_cast<Vector6d> (ATA.inverse () * ATb);
+  Vector6d x = static_cast<Vector6d> (ATA.selfadjointView<Eigen::Upper> ().ldlt ().solve (ATb));
```
ATA is a least-squares matrix and therefore symmetric positive semidefinite. If it's strictly positive definite, the Cholesky decomposition will be more efficient than inverting the matrix because it allows you to take advantage of the symmetric structure. If its semidefinite and has zero eigenvalues, both the decomposition and the eigenvalues will theoretically fail. However, this is probably such a rare case that the additional cost of a rank-revealing decomposition is not worth it.

matthewdcong commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -186,34 +167,27 @@ estimateRigidTransformation (ConstCloudIterator<PointSource>& source_it, ConstCl
     double b = nx*sz - nz*sx; 
```
Sure, happy to change.

matthewdcong commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -186,34 +167,27 @@ estimateRigidTransformation (ConstCloudIterator<PointSource>& source_it, ConstCl
-    ATA.coeffRef (28) += ny * ny;
-    ATA.coeffRef (29) += ny * nz;
-    ATA.coeffRef (35) += nz * nz;
+    U.coeffRef (0, 0) += a * a;
```
I've rewritten this using matrix vector multiplies and the `Eigen::selfadjointView` which eliminates the need to explicitly loop over indices. Overall, I think the result is much more readable and concise. Let me know what you think.

kunaltyagi replied:
I think you forgot to push 😄

matthewdcong replied:
Yeah, sorry one sec. Needed to rebuild from scratch 😅

matthewdcong replied:
OK, sorry about that. Just pushed the change!

matthewdcong commented on registration/include/pcl/registration/transformation_estimation_symmetric_point_to_plane_lls.h:
```diff
@@ -0,0 +1,150 @@
+
+        using Matrix4 = typename TransformationEstimation<PointSource, PointTarget, Scalar>::Matrix4;
+
+        using Vector6d = Eigen::Matrix<double, 6, 1>;
```
Done

matthewdcong commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,207 @@
+  rotation_matrix = Eigen::AngleAxis<Scalar> (parameters (2), Eigen::Matrix<Scalar, 3, 1>::UnitZ ()) * 
+                    Eigen::AngleAxis<Scalar> (parameters (1), Eigen::Matrix<Scalar, 3, 1>::UnitY ()) *
+                    Eigen::AngleAxis<Scalar> (parameters (0), Eigen::Matrix<Scalar, 3, 1>::UnitX ());
+  Eigen::Matrix<Scalar, 3, 1> translation_vector;
```
Sounds good

matthewdcong commented on common/include/pcl/common/point_tests.h:
```diff
@@ -101,4 +101,13 @@ namespace pcl
+
+  // specification for pcl::PointNormal
+  template <> inline bool
+  isFinite<pcl::PointNormal> (const pcl::PointNormal &p)
```
Good point, I removed this part of the change and switched to using Eigen::isFinite

matthewdcong commented:
Thanks for the comments! I've addressed them in my latest change. I've also incorporated the improved objective linearization proposed in the paper.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    Vector6 v;
+    v << (s + d).cross (n), n;
```
This looks so elegant. Kudos

matthewdcong replied:
Thank you!

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -116,90 +116,57 @@ estimateRigidTransformation (const pcl::PointCloud<PointSource> &cloud_src,
 
   // Solve A*x = b
-  Vector6d x = static_cast<Vector6d> (ATA.selfadjointView<Eigen::Upper> ().ldlt ().solve (ATb));
+  auto x = static_cast<Vector6> (M.ldlt ().solve (ATb));
```
`const` for x

matthewdcong replied:
Fixed

matthewdcong commented:
Yeah, and correspondingly in `constructTransformationMatrix`.

L162 would be` s.cross(n)` with the transformation matrix equal to `R * T` in a non-symmetric linearization. The symmetric version has `(s+d).cross(n)` and the transformation matrix equal to `R*T*R`.

kunaltyagi commented:
Any reason for the following test to fail with an off-by-1 error on MacOS?

```
98: [ RUN      ] PCL.Octree_Pointcloud_Neighbours_Within_Radius_Search
98: /Users/vsts/agent/2.155.1/work/1/s/test/search/test_octree.cpp:330: Failure
98: Expected equality of these values:
98:   cloudNWRRadius.size()
98:     Which is: 547
98:   cloudSearchBruteforce.size()
98:     Which is: 548
98: [  FAILED  ] PCL.Octree_Pointcloud_Neighbours_Within_Radius_Search (25 ms)
98: [  FAILED  ] PCL.Octree_Pointcloud_Neighbours_Within_Radius_Search
98: 
98:  1 FAILED TEST
 98/106 Test  #98: octree_search ..........................***Failed    0.05 sec
```

matthewdcong commented:
Strange...I don’t think my changes affect that portion of the library. I’m on Mac OSX and everything passed on my end.

taketwo commented:
Let me read the paper and get back to you with comments.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -116,90 +116,57 @@ estimateRigidTransformation (const pcl::PointCloud<PointSource> &cloud_src,
   {
-    if (!isFinite(*source_it) ||
-        !isFinite(*target_it))
+    Vector3 s (source_it->x, source_it->y, source_it->z);
```
I notice that the values are not mean subtracted. Will that impact convergence or numerical stability as hinted in the paper? (Eq 10 is EM for mean-subtracted set of points, rest seems similar)

taketwo replied:
Fair point. The question is, who and when should take care of demeaning. I don't see an efficient way to perform it within this class. And generally, mean subtraction is a good idea for every objective, not only symmetric. So I feel that there should be a general mechanism to do this though it's outside of the scope of this PR then.

kunaltyagi replied:
There's no need to perform demeaning. Find the mean in a separate pass (or get it as parameter, defaulted to `Eigen3X::Zero()` and `const auto s = Vector3(...) - mean;` No need to add the mean

matthewdcong replied:
I've subtracted the mean from the point clouds in order to improve numerical stability/convergence.

taketwo commented on registration/CMakeLists.txt:
```diff
@@ -166,6 +168,7 @@ set(srcs
   src/transformation_estimation_point_to_plane_weighted.cpp
   src/transformation_estimation_point_to_plane_lls.cpp
   src/transformation_estimation_point_to_plane_lls_weighted.cpp
+  src/transformation_estimation_symmetric_point_to_plane_lls.cpp
```
Please get rid of this file entirely.

matthewdcong replied:
Removed

taketwo commented on registration/include/pcl/registration/icp.h:
```diff
@@ -331,6 +332,54 @@ namespace pcl
+    * \ingroup registration
+    */
+  template <typename PointSource, typename PointTarget, typename Scalar = float>
+  class SymmetricIterativeClosestPointWithNormals : public IterativeClosestPoint<PointSource, PointTarget, Scalar>
```
I wonder if we need to introduce an entire new class for this. What do you think about adding an option "use symmetric objective" to the 'IterativeClosestPointWithNormals' class?

kunaltyagi replied:
Or more generally:

```Vector3 n = normal_weighing_function(source_it->getNormalVector3fMap(), target_it->getNormalVector3fMap(), weight_source, weight_target);```

We can take this as lambda input and simply provide some default lambdas who's behavior boils down to:
* weight_source = 1, weight_target = 0: obv
* weight_source = 1, weight_target = 1: symetric
* weight_source = 0, weight_target = 1:: point-to-plane weighted

matthewdcong replied:
I've removed the new class and added the use symmetric objective option.

matthewdcong replied:
Thanks for the suggestion, but could we please defer the weighting lambda to another PR? I think it could help with unifying weighting among the existing ICP implementations as well (e.g. point to plane vs. point to plane weighted which currently exist as separate classes). Thus, I feel that it's a bit out of scope for this feature PR.

taketwo replied:
> Thanks for the suggestion, but could we please defer the weighting lambda to another PR?

Certainly. I've already seen way too many PRs that were sidetracked by endless refactoring opportunities :laughing:

taketwo commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,174 @@
+ *
+ */
+
+#ifndef PCL_REGISTRATION_TRANSFORMATION_ESTIMATION_SYMMETRIC_POINT_TO_PLANE_LLS_HPP_
```
Please use `#pragma once`. (We still have some old-style guards here and there, but new code should use pragmas.)

matthewdcong replied:
Done

taketwo commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,174 @@
+  const auto nr_points = cloud_src.points.size ();
+  if (cloud_tgt.points.size () != nr_points)
+  {
+    PCL_ERROR ("[pcl::TransformationEstimationSymmetricPointToPlaneLLS::estimateRigidTransformation] Number or points in source (%lu) differs than target (%lu)!\n", nr_points, cloud_tgt.points.size ());
```
```suggestion
    PCL_ERROR ("[pcl::TransformationEstimationSymmetricPointToPlaneLLS::estimateRigidTransformation] Number or points in source (%lu) differs from target (%lu)!\n", nr_points, cloud_tgt.points.size ());
```

matthewdcong replied:
Good catch

taketwo commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,174 @@
+  for (; source_it.isValid () && target_it.isValid (); ++source_it, ++target_it)
+  {
+    Vector3 s (source_it->x, source_it->y, source_it->z);
+    Vector3 d (target_it->x, target_it->y, target_it->z);
```
I propose to name variables the same as in paper, i.e. `s -> p`, `d -> q`.

matthewdcong replied:
done

taketwo commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,174 @@
+  {
+    Vector3 s (source_it->x, source_it->y, source_it->z);
+    Vector3 d (target_it->x, target_it->y, target_it->z);
+    Vector3 n (source_it->normal[0] + target_it->normal[0],
```
```cpp
Vector3 n = source_it->getNormalVector3fMap() + target_it->getNormalVector3fMap();
```

matthewdcong replied:
Much more concise/convenient, thanks!

taketwo commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  // Solve A*x = b
+  const auto x = static_cast<Vector6> (M.ldlt ().solve (ATb));
```
This looks strange to me. Wouldn't `const Vector6 x = M.ldlt ().solve (ATb)` work?

matthewdcong replied:
Yeah, that works

taketwo commented on registration/include/pcl/registration/transformation_estimation_symmetric_point_to_plane_lls.h:
```diff
@@ -0,0 +1,150 @@
+      *
+      * \note The class is templated on the source and target point types as well as on the output scalar of the
+      * transformation matrix (i.e., float or double). Default: float.
+      * \author Michael Dixon
```
Don't be shy :wink:

kunaltyagi commented:
>>     only makes a difference to the outlier rejection strategy described in Section 4.3.
> if this is needed in our case at all.

From section 4.3:
> we adopt a relatively simple outlier-rejection strategy that excludes matches having normals with negative dot product, and pairs with (point-to-point) distance greater than 2.5 times the standard deviation of distances at each iteration [Masuda et al.1996]. (The standard devia-tion is estimated robustly as 1.4826 times the median distance.)

There's no outlier rejection in our case (or the explicit iterations) so I feel that's not required. Having said that, I do not know if this would improve performance.

Intuitively, the correct choice should improve the objective (eg: normal pointed inwards vs normal pointed outwards) since at optimized situation, both have different values of `n_q_i + n_p_i` (2 versus 0) but mathematically, I can't justify inverting one if the dot product is negative with the same intuition (which make this last paragraph useless)

matthewdcong commented:
Thank you both for the feedback! I've taken a pass at addressing this round of comments in my latest two commits.

kunaltyagi commented on registration/include/pcl/registration/impl/transformation_estimation_symmetric_point_to_plane_lls.hpp:
```diff
@@ -0,0 +1,191 @@
+  const Eigen::Translation<Scalar, 3> translation (parameters (3), parameters (4), parameters (5));
+  const Eigen::Translation<Scalar, 3> source_translation (-source_centroid.template head<3> ());
+  const Eigen::Translation<Scalar, 3> target_translation (target_centroid.template head<3> ());
+  const Eigen::Transform<Scalar, 3, Eigen::Affine> transform = target_translation *
```
Is the rotation matrix correct? Wouldn't one of the rotations be inverse?

matthewdcong commented:
Yes, the transformation matrix is correct. The rotation applied to `q` is inverted in the formulation of the objective function. See the text in the paper immediately following Equation 5. This inversion in the objective function leads to the sign change in the cross product as compared to the dot product in the linearized formulation.

taketwo commented:
## Demeaning

> The question is, who and when should take care of demeaning. I don't see an efficient way to perform it within this class. 

Let me elaborate on this. The class in question implements a single alignment estimation iteration. Your latest commit adds mean computation right into the body of the iteration. I don't think it's efficient (or indeed needed) to compute mean in every iteration.

## Oriented normals

> There's no outlier rejection in our case (or the explicit iterations) so I feel that's not required. 

Yes, not inside this class. But outside in the ICP scheme, there is outlier rejection.

The thing is, in general case, we can not assume that our input normals are oriented. This is because normal estimation in 3D clouds is ambiguous in direction sign. Therefore, it may happen that a point from the source cloud and a point from the target cloud that are a perfect match, have opposite estimated normals. What the footnote says is, "if normals are not oriented, then we should be optimistic and assume that they are pointing in the same direction".

kunaltyagi commented:
The mean subtraction can be removed from the class since this is something that *can be done externally* before iterating in the class and then the result applied. It'll require a copy of the clouds but it woks. For now a note to user might be sufficient.

For the future, we should track this in an issue since it might require refactoring the code to allow a change of API to either supply a mean, or compute it internally in the class and expose it (to prevent wasting compute).

As for normal computation, I can see it implemented (tracked by the same issue) in one of the following ways: 
* Per cloud: a setting similar to symmetric objective setting for flipping all normals from one cloud (sounds weird. normal can be flipped locally)
* Per point: an additional input `vector<enum with entries -1 and 1>` for multiplying per point (or index)  
* Mixed: using lambda and allowing the user to weigh the normals as they wish. Lambda inputs: normal1, normal2, index/iterator

taketwo commented:
OK, I think we have an agreement regarding mean subtraction. For this PR, it's sufficient to have a note in the documentation.

Regarding normal flipping, I either don't understand your suggestion, or we are talking about different things. My point is: normal estimation algorithm produces normals that have arbitrary direction signs. For example, two neighboring points sampled from the same planar surface may end up having normals pointing in opposite directions. What's more, without additional "semantic" knowledge about the point cloud it's impossible to disambiguate the sign, i.e. tell which direction is "correct", and flip normals that are pointing in "wrong" direction. With this in mind, I don't understand your proposals. Option 1 does not work because it's not that all points need to be flipped, only some. Option 2 is closer, but still, the user can not know which points have to be flipped. Option 3 has the same problem.

To expand more on my understanding of the footnote, I imagine that on line 167, we should check the dot product of normals, and if it is negative subtract them instead of adding.

---

Apart from finishing the discussion above, I would love to see a bit more documentation added for this PR. Specifically, update the documentation for `IterativeClosestPointWithNormals` explain that it implements two objectives now and that this can be controlled via the setter. Further, it would be great to add a brief discussion of the benefits of the symmetric objective. (Simply copy-paste the claims from the paper.) The end goal is to provide users with (at least some) hints what they can expect from enabling the symmetric option.

matthewdcong commented:
Thanks for the feedback! Sorry @kunaltyagi, I think I'm going to have to agree with @taketwo on this one. From a numerical standpoint, I think checking the dot product of the normals and adding/subtracting them based on the sign of the dot product is more robust. Consider the case of two planes with identical positions but opposite normals. Without normal flipping, the energy of this configuration is 0 _regardless_ of the positions of the points and as a result any transformation minimizes the energy. In other words, adding normals with opposite directions has removed the minimum energy. This actually leads to the least squares linearization breaking as well because the normal equations matrix becomes all zeros. Normal flipping based on the dot product solves this problem.

Obviously, this is an extreme and somewhat contrived example, but if you consider complicated point clouds with potentially opposite normals, this can manifest as more local minima in the energy which leads to poor convergence.

matthewdcong commented:
I've address the latest round of reviews:

- Added normal flipping based on dot product
- Removed subtraction of the mean and added a note to the documentation.
- Added documentation on newly added symmetric objective.

Thanks again for all of the feedback and discussion!

kunaltyagi commented:
Sorry, please let me explain the points I made logical jumps on. I often make the mistake of making statements without putting them in context. Apologies for that.

> additional "semantic" knowledge about the point cloud it's impossible to disambiguate the sign

I made an assumption that the user might have other tools to disambiguate the sign or strategies to deal with the flipping. Merely exposing the fact that "the way normals interact with each other is complex and the default behavior is ... <this> ..." allows the API to warn/inform the users rather than the documentation.

IMO, the obvious ways were 3, out of which one (first one) was within the scope of the PR but was a global switch. The other two, using a vector per point or a lambda passed by the user (with default behavior of simply adding them) require change in API that would be better done in another PR (possibly same one with the API change to add something regarding mean)

> we should check the dot product of normals, and if it is negative subtract them instead of adding
> I think ... is more robust

I was never against this, just that in this algorithm there isn't sufficient information, which the user might have access to. I do have a concern with this (but the concern isn't for this PR) since this is a local strategy as it ignores the global information (which it can't access). The normal direction being ignored can have effects on the computed alignment since the computed correspondence as well as the normals can be noisy. Flipping the normals is equivalent to solving the equation with a misalignment of `180-theta` in the normals instead of `theta` which essentially changes the local features in the equation.

Having said that, I agree that subtracting when dot product is negative is a good default behavior since (as pointed above) it tries to keep the energy high instead of being optimistic and adding which can bring the energy to 0.

The author actually provides code and data on the website for testing any hypothesis/improvements we have. Please do check the [author's original implementation](https://gitlab.com/kunaltyagi/trimesh2/blob/master/libsrc/ICP.cc#L201) which I've uploaded. It may contain things we wish to add in the future.

I'll be busy this weekend so I might not be able to respond till Tuesday. In that case, please go ahead with what seems natural. I can always make an issue out of it later 😆

matthewdcong commented:
Thanks for the explanation. I'd be happy to add an option to `IterativeClosestPoint` similar to the `setUseSymmetricObjective` and propagate that down to the symmetric ICP implementation. As you mentioned, that would allow the API to implicitly warn the user.

Does that sound good to everyone?

matthewdcong commented:
Added option to toggle normal flipping.

matthewdcong commented:
I think squashing everything into a single commit makes sense. Is there anything I need to do on my end for the squashing? Or is it squashed as a part of the merge process?

taketwo commented:
> Or is it squashed as a part of the merge process?

Yep, we can do this on our side. Let's just wait for the last word from Kunal.

kunaltyagi commented:
Looks good!! 😄  The best outcome from this PR (apart from the core contrib) is the potential to make the ICP algorithms better using Eigen and class flags.

Maybe I missed it (since I just reviewed the changes), but did we cite the academic paper?

matthewdcong commented:
Yup, it’s referenced in the header. See the second commit in this PR.

taketwo commented:
Thanks @matthewdcong for the contribution and quick responses, and @kunaltyagi for a thorough review.
